### PR TITLE
yamitranscode: implement VBR and HRD parameter setting

### PR DESCRIPTION
**Pull request by wudping**
Implement VBR and HRD parameter setting for yamitranscode.

for [VIZ-8997](https://jira01.devtools.intel.com/browse/VIZ-8997)

Signed-off-by: wudping <dongpingx.wu@intel.com>

### CODE REVIEW

**Comment by wudping** [comment]
please help to review, thanks.

File: tests/vppoutputencode.cpp [hunk 1]
```diff
@@ -103,10 +107,15 @@ static void setEncodeParam(const SharedPtr<IVideoEncoder>& encoder,
     encVideoParams.intraPeriod = encParam->intraPeriod;
     encVideoParams.ipPeriod = encParam->ipPeriod;
     encVideoParams.rcParams.bitRate = encParam->bitRate;
+    encVideoParams.rcParams.targetPercentage = encParam->targetPercentage;
```
Comment (xuguangxin) [inline tests/vppoutputencode.cpp]: If the user did not set this we do not need set this.This approach we hide issues for the default setting. You always set the targePercentage, this applied to other fields like windowSize.

File: tests/yamitranscode.cpp [hunk 2]
```diff
@@ -43,7 +43,11 @@ static void print_help(const char* app)
     printf("   -N <number of frames to encode(camera default 50), useful for camera>\n");
     printf("   -t <AVC scalability temporal layer number  (default 1)> optional\n");
     printf("   --qp <initial qp> optional\n");
-    printf("   --rcmode <CBR|CQP> optional\n");
+    printf("   --rcmode <CBR|CQP|VBR> optional\n");
+    printf("   --tpercentage <target percentage of bitrate in VBR mode, default 95, range in(50-100)> optional\n");
+    printf("   --windowsize <windows size in milliseconds> optional\n");
+    printf("   --bufferfullness <vbv initial buffer fullness in bit> optional\n");
+    printf("   --buffersize <vbv buffer size in bit> optional\n");
```
Comment (xuguangxin) [inline tests/yamitranscode.cpp]: name it ass vbv-buffer-size and vbv-buffer-fullness

File: tests/yamitranscode.cpp [hunk 3]
```diff
@@ -43,7 +43,11 @@ static void print_help(const char* app)
     printf("   -N <number of frames to encode(camera default 50), useful for camera>\n");
     printf("   -t <AVC scalability temporal layer number  (default 1)> optional\n");
     printf("   --qp <initial qp> optional\n");
-    printf("   --rcmode <CBR|CQP> optional\n");
+    printf("   --rcmode <CBR|CQP|VBR> optional\n");
+    printf("   --tpercentage <target percentage of bitrate in VBR mode, default 95, range in(50-100)> optional\n");
```
Comment (xuguangxin) [inline tests/yamitranscode.cpp]: just name it as --target-percnetage

File: tests/yamitranscode.cpp [hunk 4]
```diff
@@ -43,7 +43,11 @@ static void print_help(const char* app)
...
+    printf("   --windowsize <windows size in milliseconds> optional\n");
```
Comment (xuguangxin) [inline tests/yamitranscode.cpp]: --hrd-window-size is better? @QuPengfei , have suggestion for this?

File: tests/yamitranscode.cpp [hunk 5]
```diff
@@ -225,6 +237,26 @@ static bool processCmdLine(int argc, char *argv[], TranscodeParams& para)
                 case 22:
                     para.m_encParams.enableLowPower = true;
                     break;
+                case 23:
+                    para.m_encParams.targetPercentage = atoi(optarg);
+                    if (para.m_encParams.targetPercentage < 50) {
+                        printf("targetPercentage(%d) shoul be in [50, 100], change to 50\n", para.m_encParams.targetPercentage);
```
Comment (xuguangxin) [inline tests/yamitranscode.cpp]: should..
Adjust this in core library maybe better than here.

File: tests/yamitranscode.cpp [hunk 6]
```diff
@@ -88,30 +95,35 @@ static bool processCmdLine(int argc, char *argv[], TranscodeParams& para)
...
-        {"refmode", required_argument, NULL, 0 },
-        {"ow", required_argument, NULL, 0 },
-        {"oh", required_argument, NULL, 0 },
-        {"btl0", required_argument, NULL, 0 },
-        {"btl1", required_argument, NULL, 0 },
-        {"btl2", required_argument, NULL, 0 },
-        {"btl3", required_argument, NULL, 0 },
-        {"lowpower", no_argument, NULL, 0 },
-        {NULL, no_argument, NULL, 0 }};
+        { "help", no_argument, NULL, 'h' },
```
Comment (xuguangxin) [inline tests/yamitranscode.cpp]: could you send other commit for this change?

File: tests/vppoutputencode.cpp (hunk 1 above)
Comment (wudping) [inline tests/vppoutputencode.cpp]: okay, thanks @xuguangxin ;
how about that:
`if (encParam->targetPercentage != DEFAULT)
    encVideoParams.rcParams.targetPercentage = encParam->targetPercentage;`

File: tests/yamitranscode.cpp (hunk 3 above)
Comment (wudping) [inline tests/yamitranscode.cpp]: okay thanks @xuguangxin, I will modify. But I think of it's a little bit too long for a command parameter.

File: tests/yamitranscode.cpp (hunk 4 above)
Comment (wudping) [inline tests/yamitranscode.cpp]: okay, thanks @xuguangxin , It is with more information.

File: tests/yamitranscode.cpp (hunk 5 above)
Comment (wudping) [inline tests/yamitranscode.cpp]: okay, thanks @xuguangxin , agree.

File: tests/vppoutputencode.cpp (hunk 1 above)
Comment (wudping) [inline tests/vppoutputencode.cpp]: but libyami also assigned the target_percentage even the value is equal to the DEFAULT.
`rateControl->target_percentage = m_videoParamCommon.rcParams.targetPercentage;`
so I think I should modify the libyami's code to protect the driver's default value.

Comment (xuguangxin) [inline tests/vppoutputencode.cpp]: Yes, libyami should make sure it work well, no matter caller set the value or not

File: tests/yamitranscode.cpp (hunk 3 above)
Comment (xuguangxin) [inline tests/yamitranscode.cpp]: clearer, and do not make user guess is important to short name.

**Comment by wudping** [comment]
updated, please help to review.

**Comment by wudping** [comment]
set the default value of windowSize equal to 1000(ms). please help to review.

**Comment by wudping** [comment]
add quality level setting, please help to review.

File: tests/yamitranscode.cpp (hunk 6 above)
Comment (wudping) [inline tests/yamitranscode.cpp]: okay, it's done, Thanks @xuguangxin .

**Pull request merged**